packages/fuels-programs/src/retry.rs

```diff
@@ -0,0 +1,455 @@
+pub struct RetryConfig {
+    pub max_attempts: usize,
+    pub interval: Backoff,
+}
+
+impl RetryConfig {
+    pub fn new(max_attempts: usize, interval: Backoff) -> Self {
+        RetryConfig {
+            max_attempts,
+            interval,
+        }
+    }
+}
+
+/// Retries an asynchronous action with customizable retry behavior.
+///
+/// This function takes an asynchronous action represented by a closure `action`.
+/// The action is executed repeatedly with backoff and retry logic based on the
+/// provided `retry_config` and the `should_retry` condition.
+///
+/// The `action` closure should return a `Future` that resolves to a `Result<T, K>`,
+/// where `T` represents the success type and `K` represents the error type.
+///
+/// # Parameters
+///
+/// - `action`: The asynchronous action to be retried.
+/// - `retry_config`: A reference to the retry configuration.
+/// - `should_retry`: A closure that determines whether to retry based on the result.
+///
+/// # Return
+///
+/// Returns `Ok(T)` if the action succeeds without requiring further retries.
+/// Returns `Err(K)` if the maximum number of attempts is reached and the action
+/// still fails. If a retryable error occurs during the attempts, the error will
+/// be returned if the `should_retry` condition allows further retries.
+///
+/// # Examples
+///
+/// ```rust
+/// use std::time::Duration;
+///use fuels_core::types::errors::Error;
+///
+/// use fuels_programs::retry::{Backoff, retry, RetryConfig};
+///
+/// async fn network_request() -> Result<(), Error> {
+///     // Simulate network request here
+///     // ...
+///     // For demonstration purposes, always return an error
+///   Err(Error::InvalidData("Error".to_string()))
+/// }
+///
+/// fn main() {
+///     let retry_config = RetryConfig {
+///         max_attempts: 3,
+///         interval: Backoff::Linear(Duration::from_secs(1)),
+///     };
+///
+///     let should_retry = |result: &Result<(), Error>| {
+///         // Retry if the error is retryable
+///         result.is_err()
+///     };
+///
+///     let result = retry(network_request, &retry_config, should_retry);
+/// }
+/// ```
+pub async fn retry<Fut, T, K, ShouldRetry>(
+    mut action: impl FnMut() -> Fut,
+    retry_config: &RetryConfig,
+    should_retry: ShouldRetry,
+) -> Result<T, K>
+where
+    T: Clone + Debug,
+    Fut: Future<Output = Result<T, K>>,
+    K: Clone + 'static + From<Error>,
+    ShouldRetry: Fn(&Result<T, K>) -> bool,
+{
+    let mut last_err = None;
+
+    for attempt in 0..retry_config.max_attempts {
+        let result = action().await;
+        match result.clone() {
+            Ok(value) if !should_retry(&result) => return Ok(value),
+            Err(error) if should_retry(&result) => last_err = Some(error.clone()),
+            Err(error) => return Err(error),
+            _ => (),
+        }
+
+        tokio::time::sleep(retry_config.interval.wait_duration(attempt)).await;
+    }
+
+    last_err.map_or_else(
```

Comment:
Can we have something like if attempts == 0 -> Error on the top of the method.
IMO it's more readable that way

Reply:
Ok